### PR TITLE
[Bugfix] fail readURL on http code > 399

### PR DESCRIPTION
**Pull request by steffakasid**
if external reference returned http code > 399 should return an error

### CODE REVIEW

File: openapi3/swagger_loader_http_error_test.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+}`)
+
+	httpmock.Activate()
+	defer httpmock.DeactivateAndReset()
+
+	httpmock.RegisterResponder("GET", "http://localhost:7965/components.openapi.json",
+		httpmock.NewStringResponder(400, ""))
+	loader := NewSwaggerLoader()
+	loader.IsExternalRefsAllowed = true
+	swagger, err := loader.LoadSwaggerFromDataWithPath(spec, &url.URL{Path: "testdata/testfilename.openapi.json"})
```
Comment (fenollp) [inline openapi3/swagger_loader_http_error_test.go:44]: Could you also test `LoadSwaggerFromData`? Reuse the `loader` please.

**Review by fenollp — CHANGES_REQUESTED** [review]
(fixes https://github.com/getkin/kin-openapi/issues/343)

Looks correct, thanks! Just a few things to go through and it'll get in :)

File: openapi3/swagger_loader.go [hunk 2]
```diff
@@ -113,6 +113,9 @@ func (swaggerLoader *SwaggerLoader) readURL(location *url.URL) ([]byte, error) {
 			return nil, err
 		}
 		defer resp.Body.Close()
+		if resp.StatusCode > 399 {
+			return nil, fmt.Errorf("request returned status code %d", resp.StatusCode)
```
[the remark this answers is not in the document]
Comment (steffakasid) [reply openapi3/swagger_loader.go:117]: I had this before but found out that there is an overall error mentioning the lookup of the location failed. Therefore I removed it from the message.

File: openapi3/swagger_loader_http_error_test.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+}`)
+
+	httpmock.Activate()
```
[the remark this answers is not in the document]
Comment (steffakasid) [reply openapi3/swagger_loader_http_error_test.go:37]: Looks quite good. I'll change this later.

File: openapi3/swagger_loader_http_error_test.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	defer httpmock.DeactivateAndReset()
+
+	httpmock.RegisterResponder("GET", "http://localhost:7965/components.openapi.json",
+		httpmock.NewStringResponder(400, ""))
+	loader := NewSwaggerLoader()
+	loader.IsExternalRefsAllowed = true
+	swagger, err := loader.LoadSwaggerFromDataWithPath(spec, &url.URL{Path: "testdata/testfilename.openapi.json"})
+
+	require.Nil(t, swagger)
+	require.Error(t, err)
```
[the remark this answers is not in the document]
Comment (steffakasid) [reply openapi3/swagger_loader_http_error_test.go:47]: Ah I was looking for something like that. Usually I use assert not require. Thanks for your input.

File: openapi3/swagger_loader_http_error_test.go (hunk 1 above)
Comment (steffakasid) [reply openapi3/swagger_loader_http_error_test.go:44]: I'll take a look at this.

**Review by fenollp — APPROVED** [review]

**Pull request merged**